File: Java/Eta/ValueAdd/src/main/java/com/rtsdk/eta/valueadd/domainrep/rdm/login/LoginConnectionConfig.java
package com.rtsdk.eta.valueadd.domainrep.rdm.login;

import java.nio.ByteBuffer;
import java.util.List;

import com.rtsdk.eta.codec.Buffer;
import com.rtsdk.eta.codec.CodecFactory;
import com.rtsdk.eta.codec.CodecReturnCodes;
import com.rtsdk.eta.rdm.Login.ServerTypes;

/**
 * Connection config is representation of login response payload and contains
 * standby configuration information.
 */
public interface LoginConnectionConfig
{
    /**
     * Performs a deep copy of {@link LoginConnectionConfig} object.
     * 
     * @param destConnectionConfig ConnectionConfig object to copy this object
     *            into. It cannot be null.
     * 
     * @return UPA return value indicating success or failure of copy operation.
     */
    public int copy(LoginConnectionConfig destConnectionConfig);

    /**
     * Clears the current contents of the server info object and prepares it
     * for re-use.
     */
    public void clear();
    
    /**
     * Returns the number of standby servers.
     * 
     * @return numStandbyServers
     */
    public long numStandbyServers();

    /**
     * Sets the number of standby servers.
     *
     * @param numStandbyServers the num standby servers
     */
    public void numStandbyServers(long numStandbyServers);

    /**
     * Sets list of {@link ServerInfo} for standby connection configuration.
     *
     * @param serverInfo the server info
     */
    public void serverList(List<ServerInfo> serverInfo);

    /**
     * Returns list of {@link ServerInfo} for standby connection configuration.
     * 
     * @return serverList
     */
    public List<ServerInfo> serverList();

    /**
     * Information about available servers. A list of RDMServerInfo is
     * used by an RDMLoginRefresh to list servers available for connecting
     * to, and whether to use them as Standby servers.
     * 
     * @see ServerInfoFlags
     * @see LoginRefresh
     */
    public static class ServerInfo
    {
        private int flags;

        private long serverIndex;
        private Buffer hostName;
        private long port;
        private long loadFactor;
        private int serverType;

        private final static String eol = System.getProperty("line.separator");
        private final static String tab = "\t";
        private StringBuilder stringBuf = new StringBuilder();

        /**
         * Instantiates a new server info.
         */
        public ServerInfo()
        {
            flags = 0;
            hostName = CodecFactory.createBuffer();
            loadFactor = 65535;
            serverType = ServerTypes.STANDBY;
            serverIndex = 0;
        }

        /**
         * Sets the server info flags. Populated by {@link ServerInfoFlags}.
         *
         * @param flags the flags
         */
        public void flags(int flags)
        {
            this.flags = flags;
        }

        /**
         * Returns the server info flags. Populated by {@link ServerInfoFlags}.
         * 
         * @return flags
         */
        public int flags()
        {
            return flags;
        }

        /**
         * Performs a deep copy of {@link ServerInfo} object.
         * 
         * @param destServerInfo ServerInfo object to copy this object into. It
         *            cannot be null.
         * 
         * @return UPA return value indicating success or failure of copy
         *         operation.
         */
        public int copy(ServerInfo destServerInfo)
        {
            assert (destServerInfo != null) : "destServerInfo can not be null";

            destServerInfo.flags(flags);

            ByteBuffer byteBuffer = ByteBuffer.allocate(hostName.length());
            hostName.copy(byteBuffer);
            destServerInfo.hostName().data(byteBuffer);
            destServerInfo.port(port);
            destServerInfo.serverIndex(serverIndex);

            if (checkHasLoadFactor())
                destServerInfo.loadFactor(loadFactor);

            if (checkHasType())
                destServerInfo.serverType(serverType);

            return CodecReturnCodes.SUCCESS;
        }

        /* (non-Javadoc)
         * @see java.lang.Object#toString()
         */
        public String toString()
        {
            stringBuf.setLength(0);
            stringBuf.append(tab);
            stringBuf.append("Server:");
            stringBuf.append(eol);
            stringBuf.append(tab);
            stringBuf.append(tab);
            stringBuf.append("hostName: ");
            stringBuf.append(hostName().toString());
            stringBuf.append(eol);

            stringBuf.append(tab);
            stringBuf.append(tab);
            stringBuf.append("port: ");
            stringBuf.append(port());
            stringBuf.append(eol);

            stringBuf.append(tab);
            stringBuf.append(tab);
            stringBuf.append("serverIndex: ");
            stringBuf.append(serverIndex());
            stringBuf.append(eol);

            if (checkHasLoadFactor())
            {
                stringBuf.append(tab);
                stringBuf.append(tab);
                stringBuf.append("loadFactor: ");
                stringBuf.append(loadFactor());
                stringBuf.append(eol);
            }
            if (checkHasType())
            {
                stringBuf.append(tab);
                stringBuf.append(tab);
                stringBuf.append("serverType: ");
                stringBuf.append(serverType());
                stringBuf.append(eol);
            }

            return stringBuf.toString();
        }

        /**
         * Copy references.
         *
         * @param srcServerInfo the src server info
         * @return the int
         */
        int copyReferences(ServerInfo srcServerInfo)
        {
            assert (srcServerInfo != null) : "srcServerInfo can not be null";

            hostName().data(srcServerInfo.hostName().data(), srcServerInfo.hostName().position(), srcServerInfo.hostName().length());
            flags(srcServerInfo.flags());
            if (srcServerInfo.checkHasLoadFactor())
                loadFactor(srcServerInfo.loadFactor());
            if (srcServerInfo.checkHasType())
                serverType(srcServerInfo.serverType());
            port(srcServerInfo.port());
            serverIndex(srcServerInfo.serverIndex());

            return CodecReturnCodes.SUCCESS;
        }

        /**
         * Clears the current contents of the server info object and prepares it
         * for re-use.
         */
        public void clear()
        {
            hostName.clear();
            port = 0;
            loadFactor = 65535;
            serverType = ServerTypes.STANDBY;
            serverIndex = 0;
        }

        /**
         * Returns hostName for the server.
         * 
         * @return Server's hostName to connect to.
         */
        public Buffer hostName()
        {
            return hostName;
        }

        /**
         * Sets hostName for the server. Note that this creates garbage if
         * buffer is backed by String object.
         * 
         * @param hostName to connect to.
         */
        public void hostName(Buffer hostName)
        {
            assert (hostName != null) : "hostName can not be null";

            hostName().data(hostName.data(), hostName.position(), hostName.length());
        }

        /**
         * Returns port for the server.
         * 
         * @return Server's port to connect to.
         */
        public long port()
        {
            return port;
        }

        /**
         * Sets port for the server.
         * 
         * @param port to connect to.
         */
        public void port(long port)
        {
            this.port = port;
        }

        /**
         * Returns Server's loadFactor.
         * 
         * @return Server's loadFactor.
         */
        public long loadFactor()
        {
            return loadFactor;
        }

        /**
         * Sets Server's loadFactor.
         *
         * @param loadFactor the load factor
         */
        public void loadFactor(long loadFactor)
        {
            assert (checkHasLoadFactor());
            this.loadFactor = loadFactor;
        }

        /**
         * Applies loadFactor presence flag.
         * 
         * This flag can also be bulk-set by {@link #flags(int)}
         * 
         */
        public void applyHasLoadFactor()
        {
            flags |= ServerInfoFlags.HAS_LOAD_FACTOR;
        }

        /**
         * Checks the presence of loadFactor field.
         * 
         * This flag can also be bulk-get by {@link #flags()}
         * 
         * @return true - if loadFactor is present, false - if not.
         */
        public boolean checkHasLoadFactor()
        {
            return (flags & ServerInfoFlags.HAS_LOAD_FACTOR) != 0;
        }

        /**
         * Returns Server's serverType.
         * 
         * @return Server's serverType.
         */
        public int serverType()
        {
            return serverType;
        }

        /**
         * Sets Server's serverType.
         *
         * @param serverType the server type
         */
        public void serverType(int serverType)
        {
            assert (checkHasType());
            this.serverType = serverType;
        }

        /**
         * Applies serverType presence flag.
         * 
         * This flag can also be bulk-set by {@link #flags(int)}
         * 
         */
        public void applyHasType()
        {
            flags |= ServerInfoFlags.HAS_TYPE;
        }

        /**
         * Checks the presence of serverType field.
         * 
         * This flag can also be bulk-get by {@link #flags()}
         * 
         * @return true - if serverType is present, false - if not.
         */
        public boolean checkHasType()
        {
            return (flags & ServerInfoFlags.HAS_TYPE) != 0;
        }

        /**
         * Sets an index for the server.
         *
         * @param serverIndex the server index
         */
        public void serverIndex(long serverIndex)
        {
            this.serverIndex = serverIndex;
        }

        /**
         * Returns an index for the server.
         * 
         * @return an index for the server.
         */
        public long serverIndex()
        {
            return serverIndex;
        }
    }

    /**
     * The RDM Login Server Info Flags.
     * 
     * @see ServerInfo
     * @see LoginRefresh
     */
    public static class ServerInfoFlags
    {
        /**
         * (0x00) No flags set
         */
        public static final int NONE = 0x00;

        /**
         * (0x01) Indicates presence of the loadFactor member.
         */
        public static final int HAS_LOAD_FACTOR = 0x01;

        /**
         * (0x02) Indicates presence of the serverType member.
         */
        public static final int HAS_TYPE = 0x02;
    }
}

File: CSharp/Eta/Src/ValueAdd/Rdm/Login/LoginConnectionConfig.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022-2023 Refinitiv. All rights reserved.         --
 *|-----------------------------------------------------------------------------
 */

using System.Diagnostics;
using System.Text;

using LSEG.Eta.Codec;
using LSEG.Eta.Rdm;

using Enum = LSEG.Eta.Codec.Enum;

namespace LSEG.Eta.ValueAdd.Rdm
{
    /// <summary>
    /// Connection config is representation of login response payload
    /// and contains standby configuration information.
    /// </summary>
    sealed public class LoginConnectionConfig
    {
        private List<ServerInfo> _serverList;

        private const string eol = "\n";
        private const string tab = "\t";
        private StringBuilder stringBuf = new();

        private Vector vector = new();
        private VectorEntry vectorEntry = new();
        private ElementList elementList = new();
        private ElementEntry elementEntry = new();
        private ElementList serverElementList = new();
        private ElementEntry serverElementEntry = new();
        private Enum tmpEnum = new();
        private UInt tmpUInt = new();
        private LocalElementSetDefDb setDb = new();

        /// <summary>
        /// Sets list of <see cref="ServerInfo"/> for standby connection configuration.
        /// </summary>
        public List<ServerInfo> ServerList
        {
            get => _serverList;
            set
            {
                Debug.Assert(value != null);
                _serverList.Clear();
                _serverList.AddRange(value);
            }
        }

        /// <summary>
        /// The number of standby servers.
        /// </summary>
        public long NumStandByServers { get; set; }

        /// <summary>
        /// Login Connection Config constructor.
        /// </summary>
        public LoginConnectionConfig()
        {
            _serverList = new List<ServerInfo>();
        }

        /// <summary>
        /// Performs a deep copy of this object into <c>destConnectionConfig</c>.
        /// </summary>
        /// <param name="destConnectionConfig">LoginConnectionConfig that will be copied into.</param>
        /// <returns><see cref="CodecReturnCode"/> indicating success or failure.</returns>
        public CodecReturnCode Copy(LoginConnectionConfig destConnectionConfig)
        {
            Debug.Assert(destConnectionConfig != null);

            destConnectionConfig.NumStandByServers = NumStandByServers;

            CodecReturnCode ret;

            foreach (ServerInfo serverInfo in ServerList)
            {
                ServerInfo destServerInfo = new ServerInfo();
                ret = serverInfo.Copy(destServerInfo);
                if (ret != CodecReturnCode.SUCCESS)
                    return ret;
                destConnectionConfig.ServerList.Add(destServerInfo);
            }

            return CodecReturnCode.SUCCESS;
        }

        /// <summary>
        /// Shallow copies the information and references contained in <c>srcConnectionConfig</c> into this object.
        /// </summary>
        /// <param name="srcConnectionConfig">LoginConnectionConfig that will be copied from.</param>
        public void CopyReferences(LoginConnectionConfig srcConnectionConfig)
        {
            Debug.Assert(srcConnectionConfig != null);

            NumStandByServers = srcConnectionConfig.NumStandByServers;
            ServerList = srcConnectionConfig.ServerList;
        }

        /// <summary>
        /// Clears the current contents of the login connection config info object and prepares it for re-use.
        /// </summary>
        public void Clear()
        {
            NumStandByServers = 0;
            _serverList.Clear();
        }

        /// <summary>
        /// Encodes the connection config info using the provided <c>encodeIter</c>.
        /// </summary>
        /// <param name="encodeIter">Encode iterator that has a buffer set to encode into.</param>
        /// <returns><see cref="CodecReturnCode"/> indicating success or failure.</returns>
        public CodecReturnCode Encode(EncodeIterator encodeIter)
        {
            // Encode Element ConnectionConfig
            elementList.Clear();
            elementList.Flags = ElementListFlags.HAS_STANDARD_DATA;

            CodecReturnCode ret = elementList.EncodeInit(encodeIter, null, 0);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            elementEntry.Clear();

            elementEntry.DataType = DataTypes.VECTOR;
            elementEntry.Name = ElementNames.CONNECTION_CONFIG;

            ret = elementEntry.EncodeInit(encodeIter, 0);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            // Encode Server Entries
            vector.Clear();
            vector.ContainerType = DataTypes.ELEMENT_LIST;
            vector.Flags = VectorFlags.HAS_SUMMARY_DATA;

            ret = vector.EncodeInit(encodeIter, 0, 0);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            // Encode numStandbyServers in summary data.
            elementList.Clear();
            elementList.Flags = ElementListFlags.HAS_STANDARD_DATA;

            ret = elementList.EncodeInit(encodeIter, null, 0);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            elementEntry.Clear();

            elementEntry.DataType = DataTypes.UINT;
            elementEntry.Name = ElementNames.NUM_STANDBY_SERVERS;
            tmpUInt.Value(NumStandByServers);
            ret = elementEntry.Encode(encodeIter, tmpUInt);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }
            ret = elementList.EncodeComplete(encodeIter, true);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            ret = vector.EncodeSummaryDataComplete(encodeIter, true);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            foreach (ServerInfo serverInfo in ServerList)
            {
                vectorEntry.Clear();
                vectorEntry.Index = (uint)serverInfo.ServerIndex;
                vectorEntry.Action = VectorEntryActions.SET;
                ret = vectorEntry.EncodeInit(encodeIter, 0);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }

                // Encode Element List describing server
                elementList.Clear();
                elementList.Flags = ElementListFlags.HAS_STANDARD_DATA;

                ret = elementList.EncodeInit(encodeIter, null, 0);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
                elementEntry.Clear();

                elementEntry.DataType = DataTypes.ASCII_STRING;
                elementEntry.Name = ElementNames.HOSTNAME;
                ret = elementEntry.Encode(encodeIter, serverInfo.HostName);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }

                elementEntry.DataType = DataTypes.UINT;
                elementEntry.Name = ElementNames.PORT;
                tmpUInt.Value(serverInfo.Port);
                ret = elementEntry.Encode(encodeIter, tmpUInt);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }

                if (serverInfo.HasLoadFactor)
                {
                    elementEntry.DataType = DataTypes.UINT;
                    elementEntry.Name = ElementNames.LOAD_FACT;
                    tmpUInt.Value(serverInfo.LoadFactor);
                    ret = elementEntry.Encode(encodeIter, tmpUInt);
                    if (ret != CodecReturnCode.SUCCESS)
                    {
                        return ret;
                    }
                }
                if (serverInfo.HasType)
                {
                    elementEntry.DataType = DataTypes.ENUM;
                    elementEntry.Name = ElementNames.SERVER_TYPE;
                    tmpEnum.Value(serverInfo.ServerType);
                    ret = elementEntry.Encode(encodeIter, tmpEnum);
                    if (ret != CodecReturnCode.SUCCESS)
                    {
                        return ret;
                    }
                }

                ret = elementList.EncodeComplete(encodeIter, true);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
                ret = vectorEntry.EncodeComplete(encodeIter, true);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            // Complete
            ret = vector.EncodeComplete(encodeIter, true);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            ret = elementEntry.EncodeComplete(encodeIter, true);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            return elementList.EncodeComplete(encodeIter, true);
        }

        /// <summary>
        /// Decodes the connection config info using the provided <c>dIter</c>.  The decode iterator must be set to decode the vector containing the login connection config info.
        /// </summary>
        /// <param name="decodeIter">Decode iterator that has already decoded the initial message and is ready to decode the Login Connection Config Vector.</param>
        /// <returns><see cref="CodecReturnCode"/> indicating success or failure.</returns>
        public CodecReturnCode Decode(DecodeIterator decodeIter)
        {
            Clear();
            CodecReturnCode ret = vector.Decode(decodeIter);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            if (vector.ContainerType != DataTypes.ELEMENT_LIST)
            {
                return CodecReturnCode.FAILURE;
            }

            if (vector.CheckHasSetDefs())
            {
                setDb.Clear();
                ret = setDb.Decode(decodeIter);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            // Decode payload
            if (vector.CheckHasSummaryData())
            {
                ret = serverElementList.Decode(decodeIter, null);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }

                bool foundNumStandbyServers = false;
                // Decode each element entry in list
                while ((ret = serverElementEntry.Decode(decodeIter)) != CodecReturnCode.END_OF_CONTAINER)
                {
                    if (ret != CodecReturnCode.SUCCESS)
                    {
                        return ret;
                    }

                    if (ElementNames.NUM_STANDBY_SERVERS.Equals(serverElementEntry.Name))
                    {
                        if (serverElementEntry.DataType != DataTypes.UINT)
                            return CodecReturnCode.FAILURE;
                        ret = tmpUInt.Decode(decodeIter);
                        if (ret != CodecReturnCode.SUCCESS)
                        {
                            return ret;
                        }
                        NumStandByServers = tmpUInt.ToLong();
                        foundNumStandbyServers = true;
                    }
                }

                if (!foundNumStandbyServers)
                    return CodecReturnCode.FAILURE;
            } // end summary data Decode

            vectorEntry.Clear();
            bool foundHostName = false;
            bool foundServerPort = false;
            if ((ret = vectorEntry.Decode(decodeIter)) != CodecReturnCode.END_OF_CONTAINER)
            {
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
                do
                {
                    ret = serverElementList.Decode(decodeIter, setDb);
                    if (ret != CodecReturnCode.SUCCESS)
                    {
                        return ret;
                    }

                    ServerInfo serverInfo = new ServerInfo();
                    ServerList.Add(serverInfo);
                    serverInfo.ServerIndex = vectorEntry.Index;

                    // Decode each element entry in list
                    while ((ret = serverElementEntry.Decode(decodeIter)) != CodecReturnCode.END_OF_CONTAINER)
                    {
                        if (ret != CodecReturnCode.SUCCESS)
                        {
                            return ret;
                        }

                        if (serverElementEntry.Name.Equals(ElementNames.HOSTNAME))
                        {
                            foundHostName = true;
                            serverInfo.HostName = serverElementEntry.EncodedData;
                        }

                        if (serverElementEntry.Name.Equals(ElementNames.PORT))
                        {
                            ret = tmpUInt.Decode(decodeIter);
                            if (ret != CodecReturnCode.SUCCESS)
                            {
                                return ret;
                            }
                            foundServerPort = true;
                            serverInfo.Port = tmpUInt.ToLong();
                        }

                        if (serverElementEntry.Name.Equals(ElementNames.LOAD_FACT))
                        {
                            ret = tmpUInt.Decode(decodeIter);
                            if (ret != CodecReturnCode.SUCCESS)
                            {
                                return ret;
                            }
                            serverInfo.HasLoadFactor = true;
                            serverInfo.LoadFactor = tmpUInt.ToLong();
                        }

                        if (serverElementEntry.Name.Equals(ElementNames.SERVER_TYPE))
                        {
                            ret = tmpEnum.Decode(decodeIter);
                            if (ret != CodecReturnCode.SUCCESS)
                            {
                                return ret;
                            }
                            serverInfo.HasType = true;
                            serverInfo.ServerType = tmpEnum.ToInt();
                        }

                    }

                    if (!foundServerPort)
                        return CodecReturnCode.FAILURE;

                    if (!foundHostName)
                        return CodecReturnCode.FAILURE;
                }
                while ((ret = vectorEntry.Decode(decodeIter)) != CodecReturnCode.END_OF_CONTAINER);
            }
            return CodecReturnCode.SUCCESS;
        }

        /// <summary>
        /// Returns a human readable string containing the connection configuration info.
        /// </summary>        
        /// <returns>String containing the string representation.</returns>
        public override string ToString()
        {
            stringBuf.Clear();

            stringBuf.Append(tab);
            stringBuf.Append("numStandbyServers: ");
            stringBuf.Append(NumStandByServers);
            stringBuf.Append(eol);

            foreach (ServerInfo serverInfo in ServerList)
            {
                stringBuf.Append(serverInfo.ToString());
            }
            stringBuf.Append(eol);

            return stringBuf.ToString();
        }
    }
}
